### PR TITLE
Extend the `utils.find_depth()` function to return depth

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR extends the function `utils.find_depth()` to return the depth (as integers) if no test assertion using `level` is given.

### CODE REVIEW

**Comment by danielhuppmann** [comment]
rebased following merge of #190

File: pyam/utils.py [hunk 1]
```diff
@@ -208,8 +208,29 @@ def style_df(df, style='heatmap'):
...
+    pipe = re.compile('\\|')
+    regexp = re.compile(_escape_regexp(s))
+
+    def _find_depth(val):
+        return len(pipe.findall(re.sub(regexp, '', val)))
+
+    # if no level test is specified, return the depth as int
+    if level is None:
+        def return_depth(val):
+            return -1 if not val.startswith(s) else _find_depth(val)
```
Comment (gidden) [inline pyam/utils.py]: Would highly suggest returning either `None` or `np.nan` or `False` instead of `-1`.

File: pyam/utils.py [hunk 2]
```diff
@@ -208,8 +208,29 @@ def style_df(df, style='heatmap'):
         return df.style.background_gradient(cmap=cm)
 
 
-def find_depth(data, s, level):
-    # determine function for finding depth level =, >=, <= |s
+def find_depth(data, s='', level=None):
+    """
+    return a list of bools asserting the depth indicated by `level`;
```
Comment (gidden) [inline pyam/utils.py]: this docstring is getting long (and a bit hard to comprehend). would you please go ahead and make it "official" with a `Params` and `Returns` section?

File: pyam/utils.py [hunk 3]
```diff
@@ -208,8 +208,29 @@ def style_df(df, style='heatmap'):
...
+
+    def _find_depth(val):
+        return len(pipe.findall(re.sub(regexp, '', val)))
+
+    # if no level test is specified, return the depth as int
+    if level is None:
+        def return_depth(val):
+            return -1 if not val.startswith(s) else _find_depth(val)
+
+        return list(map(return_depth, data))
```
Comment (gidden) [inline pyam/utils.py]: shorter and perhaps easier to comprehend later just to make this a lambda, e.g.,

```
return list(map(lambda x: _find_depth(val) if x.startswith(s) else None, data))
```

File: pyam/utils.py [hunk 4]
```diff
@@ -221,10 +242,9 @@ def find_depth(data, s, level):
     else:
         raise ValueError('Unknown level type: {}'.format(level))
 
-    # determine depth
-    pipe = re.compile('\\|')
-    regexp = str(s).replace('*', '')
-    apply_test = lambda val: test(len(pipe.findall(val.replace(regexp, ''))))
+    def apply_test(val):
```
Comment (gidden) [inline pyam/utils.py]: suggest making this a lambda similarly:

```
list(map(lambda x: test(_find_depth(x)), data))
```

Comment (danielhuppmann) [reply pyam/utils.py]: @gidden, there is a plethora of tests for `level` in `test_core.py` - do you want me to refactor them and migrate to `test_utils.py`?

**Comment by danielhuppmann** [comment]
@gidden, there is a plethora of tests for `level` in `test_core.py` - do you want me to refactor them and migrate to `test_utils.py`?

**Comment by gidden** [comment]
Ah, I was looking for tests calling `find_depth(..., level=...)`. If I missed them apologies! But all tests for a specific function should be in a single file if they aren't already.

File: pyam/utils.py (hunk 4 above)
Comment (danielhuppmann) [reply pyam/utils.py]: done

File: pyam/utils.py (hunk 3 above)
Comment (danielhuppmann) [reply pyam/utils.py]: done

File: pyam/utils.py (hunk 2 above)
Comment (danielhuppmann) [reply pyam/utils.py]: done

File: pyam/utils.py (hunk 1 above)
Comment (danielhuppmann) [reply pyam/utils.py]: refactor to `None`

**Comment by danielhuppmann** [comment]
Thanks @gidden for the comments and suggestion for improvements - much nicer now! And migrating the unit tests made me realise that there were some unexpected corner cases which are resolved now.

**Comment by gidden** [comment]
Hi @danielhuppmann I am personally in favor of named lambdas (especially in
this context). If you agree, can you add E731 to
https://github.com/IAMconsortium/pyam/blob/master/.stickler.yml#L5

File: pyam/utils.py [hunk 5]
```diff
@@ -208,24 +208,46 @@ def style_df(df, style='heatmap'):
...
+    s = s.rstrip('*')
+    regexp = re.compile('^' + _escape_regexp(s))
+
+    def _pipe(val):
+        return len(re.compile('\\|').findall(re.sub(regexp, '', val))) \
+            if regexp.match(val) else None
+
+    # if no level test is specified, return the depth as int
+    if level is None:
+        return list(map(lambda x: _pipe(x), data))
```
Comment (gidden) [inline pyam/utils.py]: sorry if I already commented here, but this can be simply: 

```
return list(map(_pipe, data))
```

File: pyam/utils.py [hunk 6]
```diff
@@ -208,24 +208,46 @@ def style_df(df, style='heatmap'):
...
     else:
-        raise ValueError('Unknown level type: {}'.format(level))
+        raise ValueError('Unknown level type: `{}`'.format(level))
 
-    # determine depth
-    pipe = re.compile('\\|')
-    regexp = str(s).replace('*', '')
-    apply_test = lambda val: test(len(pipe.findall(val.replace(regexp, ''))))
-    return list(map(apply_test, data))
+    return list(map(lambda x: test(_pipe(x)), data))
```
Comment (gidden) [inline pyam/utils.py]: probably cleaner to include the `_pipe` calls above in lambda defs and just have here:

```
return list(map(test, data))
```

File: pyam/utils.py [hunk 7]
```diff
@@ -208,24 +208,46 @@ def style_df(df, style='heatmap'):
...
+    level : int or str, default None
+        if None, return depth (number of `|`); else, return list of booleans
+        whether depth satisfies the condition (equality if `level` is int,
+        >= if `.+`,  <= if `.-`)
+    """
+    # remove wildcard as last character from string, find depth
+    s = s.rstrip('*')
+    regexp = re.compile('^' + _escape_regexp(s))
+
+    def _pipe(val):
```
Comment (gidden) [inline pyam/utils.py]: and maybe a more meaningful name here? perhaps `find_n_pipes`?

Comment (gidden) [reply pyam/utils.py]: Or just `_n_pipes`?

File: tests/test_utils.py [hunk 8]
```diff
@@ -3,102 +3,145 @@
...
+
+
+def test_find_depth():
+    obs = utils.find_depth(TEST_VARS)
+    assert obs == [0, 1, 2]
+
+
+def test_find_depth_with_str():
+    data = pd.Series(['foo', 'foo|bar|baz', 'bar|baz', 'bar|baz|foo'])
+    obs = utils.find_depth(data, 'bar')
```
Comment (gidden) [inline tests/test_utils.py]: for this (and below) tests, can you please also add explicit kwargs. Without them, it's hard to know which arguments are actually being used (especially when you look months later). So for this, it would be

```
obs = utils.find_depth(data, s='bar')
```

Sorry for so many comments, but I think I'm done!

File: pyam/utils.py (hunk 6 above)
Comment (danielhuppmann) [reply pyam/utils.py]: done

File: pyam/utils.py (hunk 5 above)
Comment (danielhuppmann) [reply pyam/utils.py]: done

**Comment by danielhuppmann** [comment]
Now good to go @gidden! Thanks for the help to make that code so smart!
